# Matrizes, Arrays, Tensores

## Referências

- Documentação oficial de Tensores do PyTorch
    http://pytorch.org/docs/master/tensors.html
- PyTorch para usuários NumPy:
    https://github.com/torch/torch7/wiki/Torch-for-Numpy-users

## NumPy array

In [2]:
import numpy as np

In [41]:
a = np.array([[2., 8., 3.],
              [0.,-1., 5.]])
a

array([[ 2.,  8.,  3.],
       [ 0., -1.,  5.]])

In [42]:
a.shape

(2, 3)

In [43]:
a.dtype

dtype('float64')

## PyTorch tensor

Os tensores do PyTorch só podem ser float, float32 ou float64

In [1]:
import torch

### Convertendo NumPy array para tensor PyTorch

In [46]:
b = torch.Tensor(np.zeros((3,4)))
b


 0  0  0  0
 0  0  0  0
 0  0  0  0
[torch.FloatTensor of size 3x4]

### Criando arrays e tensores constantes

In [7]:
c = np.ones((2,4)); c

array([[ 1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.]])

In [8]:
d = torch.ones((2,4)); d


 1  1  1  1
 1  1  1  1
[torch.FloatTensor of size 2x4]

### Criando arrays e tensores aleatórios

In [9]:
e = np.random.rand(2,4); e

array([[ 0.63917462,  0.535567  ,  0.3038379 ,  0.16857778],
       [ 0.32212575,  0.27986183,  0.34148848,  0.56301773]])

In [10]:
f = torch.rand(2,4); f


 0.0096  0.5638  0.5356  0.4865
 0.1573  0.6713  0.1298  0.3537
[torch.FloatTensor of size 2x4]

### Arrays aleatórios com semente, para reproduzir mesma sequência pseudoaleatória

In [11]:
np.random.seed(1234)
e = np.random.rand(2,4);e

array([[ 0.19151945,  0.62210877,  0.43772774,  0.78535858],
       [ 0.77997581,  0.27259261,  0.27646426,  0.80187218]])

In [12]:
torch.manual_seed(1234)
f = torch.rand(2,4); f


 0.1915  0.4977  0.6221  0.8178
 0.4377  0.6121  0.7854  0.7714
[torch.FloatTensor of size 2x4]

### Torch seed is different for GPU

In [13]:
if torch.cuda.is_available():
    torch.cuda.torch.manual_seed(1234)
    g = torch.cuda.torch.rand(2,4)
    print(g)

## Conversões entre NumPy e Tensores PyTorch

### NumPy para Tensor PyTorch utilizando `.from_numpy()` - CUIDADO

Não são todos os tipos de elementos do array NumPy que podem ser convertidos 
para tensores PyTorch. Abaixo é um programa que cria uma tabela de equivalencias
entre os tipos do NumPy e os tipos do Tensor PyTorch:

In [52]:
import pandas as pd 
dtypes = [np.uint8, np.int32, np.int64, np.float32, np.float64, np.double]
table = np.empty((2, len(dtypes)),dtype=np.object)
for i,t in enumerate(dtypes):
    a = np.array([1],dtype=t)
    ta = torch.from_numpy(a)
    table[0,i] = a.dtype.name
    table[1,i] = type(ta).__name__
pd.DataFrame(table)

0          1           2            3             4             5
0       uint8      int32       int64      float32       float64       float64
1  ByteTensor  IntTensor  LongTensor  FloatTensor  DoubleTensor  DoubleTensor

### NumPy para Tensor utilizando `torch.FloatTensor()` - método recomendado

Existe uma cuidado importante a ser tomado na transformação de matrizes do NumPy para tensores PyTorch pois as funções de rede neurais do PyTorch utilizam o tipo FloatTensor e o NumPy utiliza como default o tipo float64, o que faz uma conversão automática para DoubleTensor do PyTorch e consequentemente gerando um erro.
A recomendação é utilizar o `torch.FloatTensor` para converter NumPy para tensores PyTorch:

In [15]:
a = np.ones((2,5))
a_t = torch.FloatTensor(a)
a_t


 1  1  1  1  1
 1  1  1  1  1
[torch.FloatTensor of size 2x5]

### Tensor PyTorch para array NumPy

In [50]:
ta = torch.ones(2,3)
ta


 1  1  1
 1  1  1
[torch.FloatTensor of size 2x3]

In [51]:
a = ta.numpy()
a

array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.]], dtype=float32)

## Tensor na CPU e na GPU

In [18]:
ta_cpu = torch.ones(2,3); ta_cpu


 1  1  1
 1  1  1
[torch.FloatTensor of size 2x3]

In [19]:
if torch.cuda.is_available():
    ta_gpu = ta_cpu.cuda()
    print(ta_gpu)

## Operações em tensores

### criação de tensor e visualização do seu shape

In [20]:
a = torch.eye(4); a


 1  0  0  0
 0  1  0  0
 0  0  1  0
 0  0  0  1
[torch.FloatTensor of size 4x4]

In [21]:
a.size()

torch.Size([4, 4])

### Reshape é feito com `view` em PyTorch

In [22]:
b = a.view(2,8); b


    1     0     0     0     0     1     0     0
    0     0     1     0     0     0     0     1
[torch.FloatTensor of size 2x8]

Aqui é um exemplo criando um tensor unidimensional sequencial de 0 a 23 e em seguida uma reshape para
que o tensor fique com 4 linhas e 6 colunas

In [23]:
a = torch.arange(0,24).view(4,6);a


  0   1   2   3   4   5
  6   7   8   9  10  11
 12  13  14  15  16  17
 18  19  20  21  22  23
[torch.FloatTensor of size 4x6]

### Adição elemento por elemento

#### usando operadores

In [24]:
c = a + a; c


  0   2   4   6   8  10
 12  14  16  18  20  22
 24  26  28  30  32  34
 36  38  40  42  44  46
[torch.FloatTensor of size 4x6]

In [25]:
d = a - c ; d


  0  -1  -2  -3  -4  -5
 -6  -7  -8  -9 -10 -11
-12 -13 -14 -15 -16 -17
-18 -19 -20 -21 -22 -23
[torch.FloatTensor of size 4x6]

#### forma funcional

In [26]:
d = a.sub(c); d


  0  -1  -2  -3  -4  -5
 -6  -7  -8  -9 -10 -11
-12 -13 -14 -15 -16 -17
-18 -19 -20 -21 -22 -23
[torch.FloatTensor of size 4x6]

#### Operação in-place

In [27]:
a.sub_(c); a


  0  -1  -2  -3  -4  -5
 -6  -7  -8  -9 -10 -11
-12 -13 -14 -15 -16 -17
-18 -19 -20 -21 -22 -23
[torch.FloatTensor of size 4x6]

### Multiplicação elemento por elemento

In [28]:
d = a * c; d 


    0    -2    -8   -18   -32   -50
  -72   -98  -128  -162  -200  -242
 -288  -338  -392  -450  -512  -578
 -648  -722  -800  -882  -968 -1058
[torch.FloatTensor of size 4x6]

In [29]:
d = a.mul(c); d


    0    -2    -8   -18   -32   -50
  -72   -98  -128  -162  -200  -242
 -288  -338  -392  -450  -512  -578
 -648  -722  -800  -882  -968 -1058
[torch.FloatTensor of size 4x6]

In [30]:
a.mul_(c); a


    0    -2    -8   -18   -32   -50
  -72   -98  -128  -162  -200  -242
 -288  -338  -392  -450  -512  -578
 -648  -722  -800  -882  -968 -1058
[torch.FloatTensor of size 4x6]

### Média em tensores

In [31]:
a = torch.arange(0,24).view(4,6); a


  0   1   2   3   4   5
  6   7   8   9  10  11
 12  13  14  15  16  17
 18  19  20  21  22  23
[torch.FloatTensor of size 4x6]

In [32]:
u = a.mean(); u

11.5

In [33]:
uu = a.sum()/a.nelement(); uu

11.5

### Média com redução de eixo 

In [34]:
u_row = a.mean(dim=1); u_row


  2.5000
  8.5000
 14.5000
 20.5000
[torch.FloatTensor of size 4]

In [35]:
u_col = a.mean(dim=0); u_col


  9
 10
 11
 12
 13
 14
[torch.FloatTensor of size 6]

### Desvio padrão

In [36]:
std = a.std(); std

7.0710678118654755

In [37]:
std_row = a.std(dim=1); std_row


 1.8708
 1.8708
 1.8708
 1.8708
[torch.FloatTensor of size 4]

## Comparação speedup CPU e GPU

In [6]:
a_numpy_cpu = np.ones((1000,1000))
%timeit b = 2 * a_numpy_cpu
a_torch_cpu = torch.ones(1000,1000)
%timeit b = 2 * a_torch_cpu

100 loops, best of 3: 2.7 ms per loop
1000 loops, best of 3: 449 µs per loop


In [39]:
if torch.cuda.is_available():
    a_gpu = a_cpu.cuda()
    %timeit b = 2 * a_gpu

Rodando o código abaixo na GTX1080: speedup de 15,5
- 888 µs ± 43.4 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
- 57.1 µs ± 22.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Rodando no macbook:
- numpy: 1000 loops, best of 3: 449 µs per loop
- torch: 1000 loops, best of 3: 1.6 ms per loop

In [8]:
%timeit b1 = a_numpy_cpu.mean()
%timeit b2 = a_torch_cpu.mean()
if torch.cuda.is_available():
    %timeit c = a_gpu.mean()

1000 loops, best of 3: 449 µs per loop
1000 loops, best of 3: 1.6 ms per loop
